File: cpp/medium/0019_remove_nth_node_from_end_of_list.cpp
// Posted Solution: https://leetcode.com/problems/remove-nth-node-from-end-of-list/solutions/5214049/simple-iterative-c-approach-beats-100-of-times/
// https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *iter = head;
        int length = 0;
        while(iter != NULL){
            iter = iter->next;
            length++;
        }
        if(length == n){
            return head->next;
        }
        ListNode *prev = head;
        iter = head;
        while(iter != NULL){
            prev = iter;
            iter = iter->next;
            length--;
            if(length == n){
                prev->next = iter->next;
                return head;
            }
        }
        return NULL;
    }
};

File: python/medium/0019_remove_nth_node_from_end_of_list.py
# https://leetcode.com/problems/remove-nth-node-from-end-of-list/description/
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        delay = 1
        prev = None
        slow = head
        fast = head
        while fast != None and fast.next != None:
            if delay >= n:
                prev = slow
                slow = slow.next
            fast = fast.next
            delay += 1

        if delay == head or prev == None:
            return head.next
        else:
            prev.next = slow.next
            return head
